src/lib/krb5/os/locate_kdc.c

```diff
@@ -506,10 +535,188 @@ prof_locate_server(krb5_context context, const krb5_data *realm,
+ * hostname with an optional port, or an HTTPS URL.
+ * The format is krb5srv:[M]:udp|tcp|kkdcp:host
+ *
+ * Return a NULL *host_out if there's any problems parsing the URI.
+ */
+static void
+parse_uri_fields(const char *uri, k5_transport *transport_out,
+                 const char **host_out, int *master_out)
+
+{
+    k5_transport transport;
+    int master = FALSE;
+
+    *transport_out = 0;
+    *host_out = NULL;
+    *master_out = -1;
+
+    /* Confirm the scheme name. */
+    if (uri == NULL || (strncasecmp(uri, "krb5srv", 7) != 0))
+        return;
+
+    uri += 7;
+    if (*uri == '\0' || *uri != ':')
+        return;
+
+    uri++;
+    if (*uri == '\0')
+        return;
+
+    /* Look for a master flag or an empty flags field. */
+    if (*uri == 'm' || *uri == 'M') {
+        master = TRUE;
+        uri++;
+        if (*uri == '\0' || *uri != ':')
+            return;
+    } else if (*uri != ':') {
+        return;
+    }
+
+    /* Look for the transport type. */
+    uri++;
+    if (strncasecmp(uri, "udp", 3) == 0) {
+        transport = UDP;
+        uri += 3;
+    } else if (strncasecmp(uri, "tcp", 3) == 0) {
+        transport = TCP;
+        uri += 3;
+    } else if (strncasecmp(uri, "kkdcp", 5) == 0) {
+        /* Currently the only MS-KKDCP transport type is HTTPS. */
+        transport = HTTPS;
+        uri += 5;
+    } else {
+        return;
+    }
+
+    if (*uri == '\0' || *uri != ':')
+        return;
+
+    /* The rest of the URI is the host and any optional parts. */
+    *host_out = uri + 1;
+    *transport_out = transport;
+    *master_out = master;
+}
+
+/*
+ * Collect a list of servers from DNS URI records, for the requested service
+ * and transport type.  Problematic entries are skipped.
+ */
 static krb5_error_code
-dns_locate_server(krb5_context context, const krb5_data *realm,
-                  struct serverlist *serverlist, enum locate_service_type svc,
-                  k5_transport transport)
+locate_uri(const krb5_data *realm, const char *req_service,
+           struct serverlist *serverlist, k5_transport req_transport,
+           int default_port, krb5_boolean master_only)
+{
+    krb5_error_code ret;
+    k5_transport transport, host_trans;
+    struct srv_dns_entry *answers, *entry;
+    char *host;
+    const char *host_field, *path;
+    int port, def_port, master;
+
+    ret = k5_make_uri_query(realm, req_service, &answers);
+    if (ret)
+        return ret;
+
+    if (answers == NULL)
+        return 0;
+
+    for (entry = answers; entry != NULL; entry = entry->next) {
+        if (entry->host == NULL)
+            continue;
+
+        def_port = default_port;
+        path = NULL;
+
+        parse_uri_fields(entry->host, &transport, &host_field, &master);
+        if (host_field == NULL)
+            continue;
+
+        /* TCP_OR_UDP allows entries of any transport type, otherwise
+         * we're asking for a match. */
```

Comment:
This is a run-on sentence; use a semicolon instead of a comma.

(This comment also exposes some technical debt with regard to the internal use of TCP_OR_UDP in locate_srv.c.  In turn, I think we have a minor behavior problem where we are unnecessarily excluding MS-KKDCP when falling back to TCP after a truncated UDP request or reply.  Those problems predate this PR and do not need to be addressed.)